File: sdks/dart/test/conformance_test.dart
// 適合試験（Dart、段 A）。
//
// 凍結ベクタ（spec/vectors）に対して TypeScript の参照実装と同一の結果を出すことを確かめる。
// ベクタを実装に合わせて変更してはならない。実装を直す（ADR-0012）。
//
// 実行: dart test（sdks/dart で）

import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'package:test/test.dart';
import 'package:wheso_client/src/fixed.dart';
import 'package:wheso_client/src/wire.dart';

/// 凍結ベクタの位置。リポジトリ直下の spec/vectors を参照する。
dynamic readVector(String name) {
  final file = File('../../spec/vectors/$name');
  final text = file.readAsStringSync();
  return jsonDecode(text);
}

Uint8List hexToBytes(String hex) {
  final bytes = Uint8List(hex.length ~/ 2);
  for (var index = 0; index + 1 < hex.length; index += 2) {
    bytes[index ~/ 2] = int.parse(hex.substring(index, index + 2), radix: 16);
  }
  return bytes;
}

String bytesToHex(Uint8List bytes) {
  final builder = StringBuffer();
  for (final byte in bytes) {
    builder.write(byte.toRadixString(16).padLeft(2, '0'));
  }
  return builder.toString();
}

void main() {
  test('擬似乱数が凍結ベクタと一致する', () {
    final root = readVector('prng.json');
    final vectors = root['vectors'] as List<dynamic>;
    expect(vectors, isNotEmpty);
    for (final entry in vectors) {
      final seedText = entry['seed'] as String;
      // 64 bit の符号なし整数は Dart の int（64 bit 符号付き）と同じビット列で扱う。
      final seed = BigInt.parse(seedText).toSigned(64).toInt();
      final created = createPrng(seed);
      if (seed == 0) {
        expect(created.isOk, isFalse, reason: '種 0 は失敗する');
        continue;
      }
      expect(created.isOk, isTrue);
      var state = created.value!;
      final outputs = entry['outputs'] as List<dynamic>;
      for (final expected in outputs) {
        final stepped = prngNext(state);
        expect(stepped.isOk, isTrue);
        state = stepped.value!.state;
        final expectedValue = BigInt.parse(expected as String).toSigned(64).toInt();
        expect(stepped.value!.output, equals(expectedValue), reason: '出力が一致する');
      }
    }
  });

  test('メディアベクタが符号化・復号で一致する', () {
    final entries = readVector('media.json') as List<dynamic>;
    expect(entries, isNotEmpty);
    for (final entry in entries) {
      final expectedHex = entry['bytesHex'] as String;
      final message = entry['message'] as Map<String, dynamic>;
      final units = <Unit>[];
      for (final unit in message['units'] as List<dynamic>) {
        units.add(Unit(
          sequenceNumber: unit['sequenceNumber'] as int,
          captureTimestampUs: BigInt.parse(unit['captureTimestampUs'] as String).toSigned(64).toInt(),
          flags: unit['flags'] as int,
          spatialId: unit['spatialId'] as int,
          temporalId: unit['temporalId'] as int,
          payload: hexToBytes(unit['payloadHex'] as String),
        ));
      }
      final built = MediaMessage(
        channel: message['channel'] as int,
        senderId: message['senderId'] as int,
        units: units,
      );
      final encoded = encodeMediaMessage(built);
      expect(encoded.isOk, isTrue, reason: '符号化できる');
      expect(bytesToHex(encoded.value!), equals(expectedHex), reason: '${entry['name']}: バイト列が一致する');

      final decoded = decodeMediaMessage(hexToBytes(expectedHex));
      expect(decoded.isOk, isTrue, reason: '復号できる');
      expect(decoded.value!.channel, equals(built.channel));
      expect(decoded.value!.senderId, equals(built.senderId));
      expect(decoded.value!.units.length, equals(built.units.length));
      for (var index = 0; index < built.units.length; index += 1) {
        expect(decoded.value!.units[index].sequenceNumber, equals(built.units[index].sequenceNumber));
        expect(decoded.value!.units[index].captureTimestampUs, equals(built.units[index].captureTimestampUs));
        expect(decoded.value!.units[index].payload, equals(built.units[index].payload));
      }
    }
  });

  test('不正なベクタが同じエラー名で拒否される', () {
    final entries = readVector('invalid.json') as List<dynamic>;
    expect(entries, isNotEmpty);
    for (final entry in entries) {
      final bytes = hexToBytes(entry['bytesHex'] as String);
      final decoded = decodeMediaMessage(bytes);
      expect(decoded.isOk, isFalse, reason: '${entry['name']}: 受理しない');
      expect(wireErrorName(decoded.error!), equals(entry['expectedErrorCode'] as String),
          reason: '${entry['name']}: 同じエラー');
    }
  });

  test('破棄順位が凍結ベクタと一致する', () {
    final entries = readVector('drop-order.json') as List<dynamic>;
    expect(entries, isNotEmpty);
    for (final entry in entries) {
      final actual = dropPriority(entry['channel'] as int, entry['flags'] as int);
      expect(actual, equals(entry['expectedPriority']), reason: '${entry['name']}');
    }
  });

  test('勾配と閾値が規範どおりである', () {
    final rising = <int>[];
    final flat = <int>[];
    final falling = <int>[];
    for (var index = 0; index < 20; index += 1) {
      rising.add(10000 + index * 1000);
      flat.add(10000);
      falling.add(30000 - index * 1000);
    }
    expect(delaySlope(rising).numerator > 0, isTrue);
    expect(delaySlope(flat).numerator, equals(0));
    expect(delaySlope(falling).numerator < 0, isTrue);
    expect(delaySlope(rising).denominator > 0, isTrue, reason: '分母は常に正');
    expect(isDegrading(delaySlope(rising)), isTrue);
    expect(isDegrading(delaySlope(flat)), isFalse);
    expect(isRecovering(delaySlope(falling)), isTrue);
  });

  test('DISCARDABLE と整数除算が規範どおりである', () {
    expect(computeDiscardable(2, false, 0, 1), isFalse, reason: '音声は false');
    expect(computeDiscardable(1, true, 0, 3), isFalse, reason: 'キーは false');
    expect(computeDiscardable(1, false, 1, 3), isFalse, reason: '最上位でない層は false');
    expect(computeDiscardable(1, false, 2, 3), isTrue, reason: '最上位の層は true');

    expect(truncDiv(10, 3).value, equals(3));
    expect(truncDiv(-10, 3).value, equals(-3));
    expect(truncDiv(10, 0).isOk, isFalse);
    expect(truncDiv(9007199254740993, 3).isOk, isFalse);
  });
}

File: sdks/kotlin/src/test/kotlin/dev/wheso/ConformanceTest.kt
// 適合試験（Kotlin、段 A）。
//
// 凍結ベクタ（spec/vectors）に対して TypeScript の参照実装と同一の結果を出すことを確かめる。
// ベクタを実装に合わせて変更してはならない。実装を直す（ADR-0012）。
//
// 実行: gradle test（sdks/kotlin で）
package dev.wheso

import java.io.File
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

class ConformanceTest {
    private fun readVector(name: String): JsonElement {
        // 凍結ベクタはリポジトリ直下の spec/vectors にある。
        val file = File("../../spec/vectors/$name")
        return Json.parseToJsonElement(file.readText())
    }

    private fun hexToBytes(hex: String): ByteArray {
        val bytes = ByteArray(hex.length / 2)
        for (index in bytes.indices) {
            val part = hex.substring(index * 2, index * 2 + 2)
            bytes[index] = part.toInt(16).toByte()
        }
        return bytes
    }

    private fun bytesToHex(bytes: ByteArray): String {
        val builder = StringBuilder(bytes.size * 2)
        for (byte in bytes) {
            builder.append("%02x".format(byte.toInt() and 0xFF))
        }
        return builder.toString()
    }

    private fun intOf(element: JsonElement?): Long {
        val primitive = element?.jsonPrimitive ?: return 0L
        val text = primitive.content
        return text.toLongOrNull() ?: 0L
    }

    /** 64 bit の符号なし整数を、同じビット列の Long として読む。 */
    private fun unsignedOf(element: JsonElement?): Long {
        val text = element?.jsonPrimitive?.content ?: return 0L
        return text.toULongOrNull()?.toLong() ?: (text.toLongOrNull() ?: 0L)
    }

    @Test
    fun prngMatchesFrozenVectors() {
        val root = readVector("prng.json").jsonObject
        val vectors = root["vectors"]?.jsonArray ?: JsonArray(emptyList())
        assertTrue(vectors.isNotEmpty(), "ベクタが空でない")
        for (entry in vectors) {
            val record = entry.jsonObject
            val seed = unsignedOf(record["seed"])
            val created = createPrng(seed)
            if (seed == 0L) {
                assertFalse(created.isOk, "種 0 は失敗する")
                continue
            }
            assertTrue(created.isOk, "擬似乱数器を作れる")
            var state = (created as Outcome.Ok).value
            val outputs = record["outputs"]?.jsonArray ?: JsonArray(emptyList())
            for (expected in outputs) {
                val stepped = prngNext(state)
                assertTrue(stepped.isOk, "状態遷移できる")
                val step = (stepped as Outcome.Ok).value
                state = step.state
                assertEquals(unsignedOf(expected), step.output, "出力が一致する")
            }
        }
    }

    @Test
    fun mediaVectorsRoundTrip() {
        val entries = readVector("media.json").jsonArray
        assertTrue(entries.isNotEmpty())
        for (entry in entries) {
            val record = entry.jsonObject
            val name = record["name"]?.jsonPrimitive?.content ?: "(名前なし)"
            val expectedHex = record["bytesHex"]?.jsonPrimitive?.content ?: ""
            val message = record["message"]?.jsonObject ?: JsonObject(emptyMap())
            val units = mutableListOf<Unit2>()
            for (unit in message["units"]?.jsonArray ?: JsonArray(emptyList())) {
                val unitRecord = unit.jsonObject
                units.add(
                    Unit2(
                        sequenceNumber = intOf(unitRecord["sequenceNumber"]),
                        captureTimestampUs = unsignedOf(unitRecord["captureTimestampUs"]),
                        flags = intOf(unitRecord["flags"]).toInt(),
                        spatialId = intOf(unitRecord["spatialId"]).toInt(),
                        temporalId = intOf(unitRecord["temporalId"]).toInt(),
                        payload = hexToBytes(unitRecord["payloadHex"]?.jsonPrimitive?.content ?: ""),
                    ),
                )
            }
            val built = MediaMessage(
                channel = intOf(message["channel"]).toInt(),
                senderId = intOf(message["senderId"]),
                units = units,
            )
            val encoded = encodeMediaMessage(built)
            assertTrue(encoded.isOk, "$name: 符号化できる")
            assertEquals(expectedHex, bytesToHex((encoded as WireOutcome.Ok).value), "$name: バイト列が一致する")

            val decoded = decodeMediaMessage(hexToBytes(expectedHex))
            assertTrue(decoded.isOk, "$name: 復号できる")
            val value = (decoded as WireOutcome.Ok).value
            assertEquals(built.channel, value.channel)
            assertEquals(built.senderId, value.senderId)
            assertEquals(built.units, value.units, "$name: ユニットが一致する")
        }
    }

    @Test
    fun invalidVectorsRejectWithSameError() {
        val entries = readVector("invalid.json").jsonArray
        assertTrue(entries.isNotEmpty())
        for (entry in entries) {
            val record = entry.jsonObject
            val name = record["name"]?.jsonPrimitive?.content ?: "(名前なし)"
            val bytes = hexToBytes(record["bytesHex"]?.jsonPrimitive?.content ?: "")
            val expected = record["expectedErrorCode"]?.jsonPrimitive?.content ?: ""
            val decoded = decodeMediaMessage(bytes)
            assertFalse(decoded.isOk, "$name: 受理しない")
            assertEquals(expected, (decoded as WireOutcome.Err).kind.errorName, "$name: 同じエラー")
        }
    }

    @Test
    fun dropOrderMatchesFrozenVectors() {
        val entries = readVector("drop-order.json").jsonArray
        assertTrue(entries.isNotEmpty())
        for (entry in entries) {
            val record = entry.jsonObject
            val name = record["name"]?.jsonPrimitive?.content ?: "(名前なし)"
            val channel = intOf(record["channel"]).toInt()
            val flags = intOf(record["flags"]).toInt()
            val expectedElement = record["expectedPriority"]
            val actual = dropPriority(channel, flags)
            if (expectedElement == null || expectedElement is JsonNull) {
                assertEquals(null, actual, "$name: 破棄禁止")
                continue
            }
            assertEquals(intOf(expectedElement).toInt(), actual, "$name: 優先順位が一致する")
        }
    }

    @Test
    fun slopeAndThresholdsMatchSpecification() {
        val rising = (0 until 20).map { 10000L + it * 60000L }
        val flat = (0 until 20).map { 10000L }
        val falling = (0 until 20).map { 1200000L - it * 60000L }

        assertTrue(delaySlope(rising).numerator > 0)
        assertEquals(0L, delaySlope(flat).numerator)
        assertTrue(delaySlope(falling).numerator < 0)
        assertTrue(delaySlope(rising).denominator > 0, "分母は常に正")
        assertTrue(isDegrading(delaySlope(rising)))
        assertFalse(isDegrading(delaySlope(flat)))
        assertTrue(isRecovering(delaySlope(falling)))
    }

    @Test
    fun discardableAndDivisionMatchSpecification() {
        assertFalse(computeDiscardable(2, false, 0, 1), "音声は false")
        assertFalse(computeDiscardable(1, true, 0, 3), "キーは false")
        assertFalse(computeDiscardable(1, false, 1, 3), "最上位でない層は false")
        assertTrue(computeDiscardable(1, false, 2, 3), "最上位の層は true")

        assertEquals(3L, (truncDiv(10L, 3L) as Outcome.Ok).value)
        assertEquals(-3L, (truncDiv(-10L, 3L) as Outcome.Ok).value)
        assertFalse(truncDiv(10L, 0L).isOk)
        assertFalse(truncDiv(9007199254740993L, 3L).isOk)
    }
}
